Support masked_fill

EikanWang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88736
* #88482

Support `masked_fill` to address the GPT2 performance issue.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

EikanWang committed:
Support masked_fill

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -741,7 +750,16 @@ def load(self, name: str, index: sympy.Expr):
+            if V.graph.get_dtype(name) in [torch.bool, torch.uint8]:
+                g_tmp_buf = f"g_tmp_buffer_{var}"
+                nelements = codecache.pick_vec_isa().nelements()
+                self.loads.writeline(f"float {g_tmp_buf}[{nelements}] = {{0}};")
+                self.loads.writeline(
+                    f"flag_to_float({var} + {cexpr(new_index)}, {g_tmp_buf}, {nelements});"
+                )
```
Is it possible that we have two loads on the same `var`? If so, there would be name clashing? Maybe safer to put these two lines into a local scope?

EikanWang replied:
Since we have passed TorchBench, HF, and TIMM, we have not encountered this case. Even if a tensor is a view of another tensor, there should be two different tensors with different variable names.

EikanWang commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -57,3 +57,12 @@ template <typename T> void atomic_add(volatile T *addr, T offset) {
 }
+
+template <typename T>
+void flag_to_float(const T* src, float* dst, int64_t n) {
```
Sure.

EikanWang replied:
Added some comments.

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang committed:
Update on "Support masked_fill"


Support `masked_fill` to address the GPT2 performance issue.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge

EikanWang commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Inductor tests are failing with AttributeError: module 'torch._inductor.codecache' has no attribute 'valid_vec_isa_list'" -c weird
